### PR TITLE
Configure Modules, DB and Controller interfaces

### PR DESCRIPTION
**Pull request by vladislav-shkutovich**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -25,6 +25,7 @@
     "@nestjs/config": "^3.2.2",
     "@nestjs/core": "^10.0.0",
     "@nestjs/jwt": "^10.2.0",
+    "@nestjs/mapped-types": "*",
```
Comment (masontikhonov) [inline package.json:28]: Putting asterisk here opens the door to regressions; let's specify version

File: src/account/account.controller.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  ): Promise<Account> {
+    return await this.accountService.createAccount(createAccountDto)
+  }
+
+  @Get(':id')
+  async getAccountById(@Param() params: IdParamDto): Promise<Account> {
+    return await this.accountService.getAccountById(params.id)
+  }
+
+  @Get('/user/:id')
```
Comment (masontikhonov) [inline src/account/account.controller.ts:37]: `GET /accounts/user/:id` seems not valid in this context.
Better approach is either `GET /users/:userId/accounts/:accountId` or `GET /accounts/:id?userid={userId}`

File: src/category/category.controller.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  ): Promise<Category> {
+    return await this.categoryService.createCategory(createCategoryDto)
+  }
+
+  @Get(':id')
+  async getCategoryById(@Param() params: IdParamDto): Promise<Category> {
+    return await this.categoryService.getCategoryById(params.id)
+  }
+
+  @Get('/user/:id')
```
Comment (masontikhonov) [inline src/category/category.controller.ts:37]: Same comment as in `AccountController`

File: src/exchange-rate/dto/exchange-rate-query-params.dto.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  @IsOptional()
+  @IsDate()
+  @Type(() => Date)
+  validTo?: Date
+
+  @IsOptional()
+  @IsNumber()
+  @Type(() => Number)
+  limit?: boolean
+}
```
Comment (masontikhonov) [inline src/exchange-rate/dto/exchange-rate-query-params.dto.ts:22]: `validFrom` is missing.

Also — now it's allowed to not pass query params at all. What will be returned in this case?

File: src/operation/operation.controller.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+  ): Promise<Operation> {
+    return await this.operationService.createOperation(createOperationDto)
+  }
+
+  @Get(':id')
+  async getOperationById(@Param() params: IdParamDto): Promise<Operation> {
+    return await this.operationService.getOperationById(params.id)
+  }
+
+  @Get('/user/:id')
```
Comment (masontikhonov) [inline src/operation/operation.controller.ts:37]: Same comment as in `AccountController`

File: src/summary/schemas/summary.schema.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
+import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose'
+import { Types } from 'mongoose'
+
+import { USER_MODEL } from '../../common/constants/database.constants'
+import { AccountEntry, AccountEntrySchema } from './account-entry.schema'
```
Comment (masontikhonov) [inline src/summary/schemas/summary.schema.ts:5]: I would define it within the same file, as they doesn't seem to have any self value.

File: src/summary/schemas/summary.schema.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import { CurrencyEntry, CurrencyEntrySchema } from './currency-entry.schema'
```
Comment (masontikhonov) [inline src/summary/schemas/summary.schema.ts:6]: Same as above

File: src/summary/schemas/summary.schema.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+  @Prop({ type: Date, required: true })
+  from: Date
+
+  @Prop({ type: Date, required: true })
+  to: Date
+
+  @Prop({ type: Map, of: AccountEntrySchema, required: true })
+  accounts: Map<string, AccountEntry>
+
```
Comment (masontikhonov) [inline src/summary/schemas/summary.schema.ts:23]: 👍🏽 nice use of mongoose maps!

File: src/summary/summary.controller.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+import { IdParamDto } from '../common/dto/id-param.dto'
+import { SummaryQueryParamsDto } from './dto/summary-query-params.dto'
+import type { Summary } from './schemas/summary.schema'
+import { SummaryService } from './summary.service'
+
+@Controller(SUMMARIES_ROUTE)
+export class SummaryController {
+  constructor(private readonly summaryService: SummaryService) {}
+
+  @Get('/user/:id')
```
Comment (masontikhonov) [inline src/summary/summary.controller.ts:13]: Same as in `AccountConntroller`

File: src/transfer/dto/transfer-query-params.dto.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+import { Type } from 'class-transformer'
+import { IsDate, IsMongoId, IsOptional, IsString } from 'class-validator'
+import { Types } from 'mongoose'
+
+export class TransferQueryParamsDto {
+  @IsOptional()
+  @IsDate()
+  @Type(() => Date)
+  from?: Date
```
Comment (masontikhonov) [inline src/transfer/dto/transfer-query-params.dto.ts:9]: In this cintext, there is a confusion between body and query params: bot are using `from/to` keywords, but in one case it stands for Account and in other — for Date. I would rename it, while also keeping consistency across the API.

File: src/transfer/dto/transfer-query-params.dto.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+  @IsOptional()
+  @IsDate()
+  @Type(() => Date)
+  to?: Date
+
+  @IsOptional()
+  @IsMongoId()
+  @Type(() => Types.ObjectId)
+  accountId?: Types.ObjectId
```
Comment (masontikhonov) [inline src/transfer/dto/transfer-query-params.dto.ts:19]: Is it source or target account?

File: src/transfer/schemas/transfer.schema.ts [hunk 12]
```diff
@@ -0,0 +1,34 @@
+import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose'
+import { Types } from 'mongoose'
+
+import { USER_MODEL } from '../../common/constants/database.constants'
+import { TransferTarget, TransferTargetSchema } from './transfer-target.schema'
```
Comment (masontikhonov) [inline src/transfer/schemas/transfer.schema.ts:5]: Same as in `SummarySchema`

File: src/transfer/transfer.controller.ts [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  ): Promise<Transfer> {
+    return await this.transferService.createTransfer(createTransferDto)
+  }
+
+  @Get(':id')
+  async getTransferById(@Param() params: IdParamDto): Promise<Transfer> {
+    return await this.transferService.getTransferById(params.id)
+  }
+
+  @Get('/user/:id')
```
Comment (masontikhonov) [inline src/transfer/transfer.controller.ts:37]: Same as in `AccountController`

**Review by masontikhonov — COMMENTED** [review]
Also, in `UserController` the password is returned on each endpoint which is security vulnerability.

File: src/account/account.controller.ts (hunk 2 above)
Comment (vladislav-shkutovich) [reply src/account/account.controller.ts]: Fixed here and for same endpoints in [this commit](https://github.com/vladislav-shkutovich/NestJS-Expenses-App/pull/1/commits/bbcb7e8cd63f168062b07125c96379aa2503933c).

File: src/operation/operation.controller.ts (hunk 5 above)
Comment (vladislav-shkutovich) [reply src/operation/operation.controller.ts]: Fixed here and for same endpoints in [this commit](https://github.com/vladislav-shkutovich/NestJS-Expenses-App/pull/1/commits/bbcb7e8cd63f168062b07125c96379aa2503933c).

File: src/summary/summary.controller.ts (hunk 9 above)
Comment (vladislav-shkutovich) [reply src/summary/summary.controller.ts]: Fixed here and for same endpoints in [this commit](https://github.com/vladislav-shkutovich/NestJS-Expenses-App/pull/1/commits/bbcb7e8cd63f168062b07125c96379aa2503933c).

File: src/transfer/transfer.controller.ts (hunk 13 above)
Comment (vladislav-shkutovich) [reply src/transfer/transfer.controller.ts]: Fixed here and for same endpoints in [this commit](https://github.com/vladislav-shkutovich/NestJS-Expenses-App/pull/1/commits/bbcb7e8cd63f168062b07125c96379aa2503933c).

File: src/exchange-rate/dto/exchange-rate-query-params.dto.ts (hunk 4 above)
Comment (vladislav-shkutovich) [reply src/exchange-rate/dto/exchange-rate-query-params.dto.ts:22]: Oh, is this case all exchange rate documents will be returned. So, it's a mistake here.
I've made `baseCurrency` and `targetCurrency` query params as required:
```
export class ExchangeRateQueryParamsDto {
  @IsString()
  baseCurrency: string

  @IsString()
  targetCurrency: string

  @IsOptional()
  @IsDate()
  @Type(() => Date)
  validFrom?: Date

  @IsOptional()
  @IsDate()
  @Type(() => Date)
  validTo?: Date

  @IsOptional()
  @IsNumber()
  @Type(() => Number)
  limit?: boolean
}
```

File: src/summary/schemas/summary.schema.ts (hunk 7 above)
Comment (vladislav-shkutovich) [reply src/summary/schemas/summary.schema.ts]: Fixed, now it's a single file for all schema classes

File: src/transfer/dto/transfer-query-params.dto.ts (hunk 10 above)
Comment (vladislav-shkutovich) [reply src/transfer/dto/transfer-query-params.dto.ts]: Agreed. Now it's a united `dateFrom`/`dateTo` naming to the all query params.

```
  @IsOptional()
  @IsDate()
  @Type(() => Date)
  dateFrom?: Date

  @IsOptional()
  @IsDate()
  @Type(() => Date)
  dateTo?: Date
```

File: src/transfer/dto/transfer-query-params.dto.ts (hunk 11 above)
Comment (vladislav-shkutovich) [reply src/transfer/dto/transfer-query-params.dto.ts]: It was a my mistake, thanks!

Now it's like: 
```
  @IsOptional()
  @IsMongoId()
  @Type(() => Types.ObjectId)
  fromAccountId?: Types.ObjectId

  @IsOptional()
  @IsMongoId()
  @Type(() => Types.ObjectId)
  toAccountId?: Types.ObjectId

  @IsOptional()
  @IsString()
  fromCurrencyCode?: string

  @IsOptional()
  @IsString()
  toCurrencyCode?: string
```

File: src/transfer/schemas/transfer.schema.ts (hunk 12 above)
Comment (vladislav-shkutovich) [reply src/transfer/schemas/transfer.schema.ts]: Fixed, now it's a single file for all schema classes

**Comment by vladislav-shkutovich** [comment]
> Also, in `UserController` the password is returned on each endpoint which is security vulnerability.

Sure, it was also missed in Todo-API project. Fixed using sanitizing helper method in [this commit](https://github.com/vladislav-shkutovich/NestJS-Expenses-App/pull/1/commits/eebe472d029199a5cf6c1bdccf25ea0d9ccb5325).

```
private sanitizeUserDocument(user: Document): UserWithoutPassword {
    return user.toObject({
      transform: function (_doc, ret) {
        delete ret.password
        return ret
      },
    })
  }

...

...return this.sanitizeUserDocument(createdUser)
```

**Review by masontikhonov — APPROVED** [review]

**Pull request merged**